lib/util.js

```diff
@@ -152,6 +152,51 @@ async function streamToBuffer(method, url, status, stream, totalSize) {
     });
 }
 
+/**
+ * Converts the content of a readable Stream into a Buffer
+ * instance. WARNING: this method will read the entirety of
+ * the stream into memory; use with extreme caution.
+ *
+ * @param {String} method HTTP method
+ * @param {String} url URL requested
+ * @param {Number} status HTTP status code of response
+ * @param {ReadableStream} stream Stream to convert.
+ * @param {number} totalSize Total size of the stream.
+ * @returns {Promise<TransferMemoryBlock>} Contents of the stream in a
+ *  pooled buffer memory block.
+ */
+async function streamToPooledBuffer(method, url, status, stream, totalSize, bufferBlockMemoryAllocator) {
+    if (!bufferBlockMemoryAllocator) {
+        return streamToBuffer(method, url, status, stream, totalSize);
+    }
+
+    return bufferBlockMemoryAllocator.obtainBuffer(totalSize).then((bufferMemoryBlock) => {
+        return new Promise((resolve, reject) => {
+            let bufferOffset = 0;
+            stream.once('error', (err) => {
+                reject(new HttpStreamError(method, url, status, err));
+            });
+
+            stream.once('end', () => {
+                if (bufferOffset !== totalSize) {
+                    reject(new HttpStreamError(method, url, status, `Unexpected number of bytes read from stream. Expected ${totalSize} but got ${bufferOffset}.`));
+                    return;
+                }
+                resolve(bufferMemoryBlock);
+            });
+
+            stream.on('data', (chunk) => {
+                if(bufferOffset >= totalSize){
+                    reject(new HttpStreamError(method, url, status, `Unexpected number of bytes written to buffer from stream. Expected ${totalSize} but got ${bufferOffset}.`));
```

Comment:
Technically valid log message, but could we indicate we got too many bytes?

Reply:
Done